# Chapitre 6 - Différentes utilisations dumachine learning avec Python (4ème partie)

## 6.7 Le deep learning avec Keras

### 6.7.2 Principe d’un réseau de neurone et première utilisation avec Keras

Le package Keras que nous avons installé est basé sur une structure extrêmement
simple à mettre en oeuvre. On commence par définir le type de réseau utilisé. Dans
cet ouvrage, il s’agira de réseaux séquentiels.

Si nous prenons les données sur les télécommunications que nous avons utilisées
précédemment, nous pouvons simplement utiliser Keras pour créer un réseau de neurones assez évolué (on suppose ici que les données ont été préparées comme
dans la première partie de ce chapitre avec Scikit-Learn) :

In [1]:
import pandas as pd
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.layers import BatchNormalization, Activation

C:\Users\s4d-asus-14\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [2]:
x=pd.read_csv("../data/x_telecom.csv",index_col=0)
y=pd.read_csv("../data/y_telecom.csv",index_col=0)

In [5]:
from sklearn.model_selection import train_test_split
# on sépare nos données en apprentissage/validation avec Scikit-Learn
x_train, x_test, y_train, y_test=train_test_split(x,y,test_size=0.2)

In [6]:
# on crée le modèle
model = Sequential()

In [7]:
# on ajoute des couches avec des fonctions d’activation
model.add(Dense(50, input_shape=(22,)))
model.add(Activation("relu"))
# cette couche permet de normaliser les données
model.add(BatchNormalization())
# cette couche va faire en sorte d’éviter l’overfitting
model.add(Dropout(0.1))
# on ajoute un autre groupe de couches
model.add(Dense(10))
model.add(Activation("relu"))
model.add(BatchNormalization())
model.add(Dropout(0.1))
# on ajoute la couche de sortie
model.add(Dense(1))
model.add(Activation('sigmoid'))


In [8]:
# on définit la fonction de perte et la fonction d’optiimsation
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=["accuracy"])

In [9]:
# on ajuste le modèle
model.fit(x_train, y_train, batch_size=100, epochs=100)

Epoch 1/100
2666/2666 [==============================] - 1s 289us/step - loss: 0.7754 - acc: 0.4839
Epoch 2/100
2666/2666 [==============================] - 0s 19us/step - loss: 0.6802 - acc: 0.5716
Epoch 3/100
2666/2666 [==============================] - 0s 17us/step - loss: 0.6475 - acc: 0.6302
Epoch 4/100
2666/2666 [==============================] - 0s 17us/step - loss: 0.6120 - acc: 0.7082
Epoch 5/100
2666/2666 [==============================] - 0s 17us/step - loss: 0.5747 - acc: 0.7536
Epoch 6/100
2666/2666 [==============================] - 0s 18us/step - loss: 0.5478 - acc: 0.7907
Epoch 7/100
2666/2666 [==============================] - 0s 18us/step - loss: 0.5267 - acc: 0.8087
Epoch 8/100
2666/2666 [==============================] - 0s 20us/step - loss: 0.4973 - acc: 0.8248
Epoch 9/100
2666/2666 [==============================] - 0s 19us/step - loss: 0.4713 - acc: 0.8428
Epoch 10/100
2666/2666 [==============================] - 0s 18us/step - loss: 0.4512 - acc: 0.8466
Epoch 11

2666/2666 [==============================] - 0s 16us/step - loss: 0.2857 - acc: 0.8905
Epoch 84/100
2666/2666 [==============================] - 0s 18us/step - loss: 0.2733 - acc: 0.8976
Epoch 85/100
2666/2666 [==============================] - 0s 17us/step - loss: 0.2792 - acc: 0.9014
Epoch 86/100
2666/2666 [==============================] - 0s 17us/step - loss: 0.2777 - acc: 0.8935
Epoch 87/100
2666/2666 [==============================] - 0s 17us/step - loss: 0.2899 - acc: 0.8890
Epoch 88/100
2666/2666 [==============================] - 0s 17us/step - loss: 0.2770 - acc: 0.8965
Epoch 89/100
2666/2666 [==============================] - 0s 15us/step - loss: 0.2736 - acc: 0.8946
Epoch 90/100
2666/2666 [==============================] - 0s 16us/step - loss: 0.2720 - acc: 0.8950
Epoch 91/100
2666/2666 [==============================] - 0s 16us/step - loss: 0.2715 - acc: 0.8965
Epoch 92/100
2666/2666 [==============================] - 0s 15us/step - loss: 0.2712 - acc: 0.8987
Epoch 93/100


In [10]:
# on calcule l’AUC pour les données de validation
from sklearn.metrics import roc_auc_score
print("AUC pour RN :", roc_auc_score(y_test,
                                     model.predict_proba(x_test, 
                                                         verbose=0).reshape(-1)))

AUC pour RN : 0.8907917293496135


Les réseaux de neurones sont des méthodes extrêmement efficaces à condition
de bien les paramétrer ce qui peut s‘avérer très difficile.

### 6.7.3 Le deep learning et les réseaux de neurones à convolutions

On va maintenant travailler sur des images. On prendra les données fashion-mnist :

In [11]:
from keras.utils import to_categorical
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.layers import Conv2D, MaxPool2D
from keras.utils import np_utils
from keras.datasets import fashion_mnist

In [12]:
# récupération des données
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

# les données sont transformées pour passer sous un format d’image
# standard avec des valeurs entre 0 et 1
x_train = (x_train/255).reshape(x_train.shape[0], 28, 28,1)
x_test = (x_test/255).reshape(x_test.shape[0], 28, 28,1)

# passage à 10 colonnes pour y(une par modalité)
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

In [13]:
# construction du modèle
model = Sequential()
model.add(Conv2D(32, (3, 3), activation = 'relu', input_shape = (28,28,1)))
model.add(Conv2D(32, (3, 3), activation = 'relu'))
model.add(MaxPool2D(pool_size=(2,2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

In [14]:
# compilation du modèle
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

In [ ]:
# ajustement du modèle
model.fit(x_train, y_train, batch_size=100, epochs=10, verbose=1)

In [ ]:
# evaluation du modèle
score = model.evaluate(x_test, y_test, verbose=0)

Nous avons appliqué deux couches de convolutions à 32 neurones avec une taille
de fenêtre de 3 par 3. Puis une couche de pooling avec une réduction sur des fenêtres
de 2 par 2. Nous obtenons les résultats suivants :

In [ ]:
# obtention de l’accuracy sur les données de validation
print("Pourcentage de bien classées :", score[1])

On voit qu’on est au-delà de 92 % bien classés sur l’échantillon de validation.
Dans notre cas, on a utilisé des images directement dans le package Keras.

Généralement, les images sont stockées dans des répertoires différents. Keras
possède de nombreux outils pour charger ces images, on peut par exemple utiliser
ImageDataGenerator :

In [ ]:
from keras.preprocessing.image import ImageDataGenerator

# on définit un générateur qui pourrait modifier nos images
# (dans ce cas on passe en float 0 à 1)
train_datagen = ImageDataGenerator(rescale=1./255)
test_datagen = ImageDataGenerator(rescale=1./255)

# on définit le dossier dans lequel se trouve les images
train_generator = train_datagen.flow_from_directory('../data/train',target_size=(150, 150),
                                                    batch_size=32,class_mode='binary')
# on définit le dossier dans lequel se trouve les images
validation_generator = test_datagen.flow_from_directory('../data/validation',
                                                        target_size=(150, 150),
                                                        batch_size=32,
                                                        class_mode='binary')

In [ ]:
# on ajuste notre modèle de deep learning avec les images
# de nos dossiers
model.fit_generator(train_generator,epochs=50,
                    validation_data=validation_generator)

Vous avez maintenant tous les outils pour vous lancer dans la construction de
réseaux de deep learning avec Python.

### 6.7.4 Aller plus loin : génération de features, transfer learning, RN, GAN …
On charge facilement un modèle avec :

In [ ]:
from keras.applications.resnet50 import ResNet50
from keras.preprocessing import image
from keras.applications.resnet50 import preprocess_input
from keras.applications.resnet50 import decode_predictions
import numpy as np

In [ ]:
# on utilise le modèle ResNet50
model = ResNet50(weights='imagenet')

In [ ]:
# on récupère une nouvelle image
img_path = '../data/elephant.png'
img = image.load_img(img_path, target_size=(224, 224))
x = image.img_to_array(img)
x = np.expand_dims(x, axis=0)
x = preprocess_input(x)

In [ ]:
# on applique le modèle
preds = model.predict(x)
print('Prédit (classe/ %):', decode_predictions(preds, top=1)[0])

On obtient donc un résultat à partir du modèle chargé. On pourra utiliser ce modèle
et ses poids dans des couches intermédiaires de notre modèle.

Dans ce cas, on a chargé un modèle et on lui a fourni une photographie d’éléphant.
Le modèle considère que c’est un éléphant d’Afrique à 87%.

Nous avons effectué un panorama approfondi de l’utilisation de Python
pour faire du machine learning, du deep learning et du traitement des
données textuelles. Python ressort comme un langage adapté à toutes
ces approches et un outil indispensable pour le data scientist.